app/services/katello/event_monitor/poller_thread.rb

```diff
@@ -0,0 +1,114 @@
+module Katello
+  module EventMonitor
+    class PollerThread
+      SLEEP_INTERVAL = 3
+      PROCESSED_COUNT_CACHE_KEY = 'katello_events_processed'.freeze
+      FAILED_COUNT_CACHE_KEY = 'katello_events_failed'.freeze
+
+      cattr_accessor :instance
+
+      def self.initialize(logger = nil)
+        self.instance ||= self.new(logger)
+      end
+
+      def self.close
+        if self.instance
+          self.instance.close
+          self.instance = nil
+        end
+        reset_status
+      end
+
+      def self.run
+        initialize
+        instance.poll_for_events
+      end
+
+      def self.status
+        {
+          processed_count: Rails.cache.fetch(PROCESSED_COUNT_CACHE_KEY) { 0 },
+          failed_count: Rails.cache.fetch(FAILED_COUNT_CACHE_KEY) { 0 },
+          queue_depth: Katello::EventQueue.queue_depth
+        }
+      end
+
+      def self.reset_status
+        Rails.cache.write(PROCESSED_COUNT_CACHE_KEY, 0)
+        Rails.cache.write(FAILED_COUNT_CACHE_KEY, 0)
+      end
+
+      def initialize(logger = nil)
+        @logger = logger || ::Foreman::Logging.logger('katello/katello_events')
+        @failed_count = 0
+        @processed_count = 0
+      end
+
+      def close
+        @logger.info("Stopping Katello Event Monitor")
+        @thread.kill if @thread
+      end
+
+      def run_event(event)
+        @logger.debug("event_queue_event: type=#{event.event_type}, object_id=#{event.object_id}")
+
+        event_instance = nil
+        begin
+          ::User.as_anonymous_admin do
+            event_instance = ::Katello::EventQueue.create_instance(event)
+            event_instance.run
+          end
+        rescue => e
+          @failed_count += 1
+          Rails.cache.write(FAILED_COUNT_CACHE_KEY, @failed_count, expires_in: 24.hours)
+
+          @logger.error("event_queue_error: type=#{event.event_type}, object_id=#{event.object_id}")
+          @logger.error(e.message)
+          @logger.error(e.backtrace.join("\n"))
+        ensure
+          if event_instance.try(:retry)
+            result = ::Katello::EventQueue.reschedule_event(event)
+            if result == :expired
+              @logger.warn("event_queue_event_expired: type=#{event.event_type} object_id=#{event.object_id}")
+            elsif !result.nil?
+              @logger.warn("event_queue_rescheduled: type=#{event.event_type} object_id=#{event.object_id}")
+            end
+          end
+          ::Katello::EventQueue.clear_events(event.event_type, event.object_id, event.created_at)
+        end
+      end
+
+      def poll_for_events
```

Comment:
@jturel I'm not sure if this is still applicable here, but we had to wrap the calls inside the thread in a `Rails.application.executor.wrap`, see https://github.com/Katello/katello/pull/7261/files because of the dev server hanging on code changes.

The discussion and some rails docs are in the PR message and I can rack my brain for more details if you need them. can you take a look and see if its still a concern here :)

Reply:
I'm glad you mentioned this! I took great care and tracked down the discussion when I came across that application executor bit. Take a look at my changes in engine.rb to see how I accomplished code reloading. The gist of it is that when code is reloaded the two event listeners are completely stopped and started again on the next request, and that seems to work perfectly. I have not seen any hangups!

Reply:
Thanks, good to hear you put some thought into it before omitting, I'll poke around and review!